scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -384,26 +397,34 @@ public boolean advance() throws IOException {
         }
 
         int keyGroupCount = 0;
+
+        // Track the canonical # buckets of each source that the key is found in.
+        // If we find it in a source with a # buckets >= the parallelism of the job,
+        // we know that it doesn't need to be re-hashed as it's already in the right bucket.
+        boolean keyFoundInSourceWithMatchingNumBuckets = false;
         while (nextKeyGroupsIt.hasNext()) {
           final Map.Entry<TupleTag, KV<byte[], Iterator<?>>> entry = nextKeyGroupsIt.next();
+
           if (keyComparator.compare(entry, minKeyEntry) == 0) {
-            if (acceptKeyGroup) {
-              int index = resultSchema.getIndex(entry.getKey());
-              @SuppressWarnings("unchecked")
-              final List<Object> values = (List<Object>) valueMap.get(index);
-              // TODO: this exhausts everything from the "lazy" iterator and can be expensive.
-              // To fix we have to make the underlying Reader range aware so that it's safe to
-              // re-iterate or stop without exhausting remaining elements in the value group.
-              entry.getValue().getValue().forEachRemaining(values::add);
-              keyGroupCount += values.size();
-            } else {
-              // Still have to exhaust iterator
-              entry.getValue().getValue().forEachRemaining(value -> {});
-            }
+            final TupleTag tupleTag = entry.getKey();
+            keyFoundInSourceWithMatchingNumBuckets |= bucketsPerSource.get(tupleTag) >= parallelism;
+            int index = resultSchema.getIndex(tupleTag);
+            @SuppressWarnings("unchecked")
+            final List<Object> values = (List<Object>) valueMap.get(index);
+            // TODO: this exhausts everything from the "lazy" iterator and can be expensive.
+            // To fix we have to make the underlying Reader range aware so that it's safe to
+            // re-iterate or stop without exhausting remaining elements in the value group.
+            entry.getValue().getValue().forEachRemaining(values::add);
+            keyGroupCount += values.size();
+
             nextKeyGroupsIt.remove();
           }
         }
 
+        final boolean acceptKeyGroup =
+            keyFoundInSourceWithMatchingNumBuckets
+                || keyGroupFilter.apply(minKeyEntry.getValue().getKey());
```

Comment:
This `keyGroupFilter.apply` is the hashing we're trying to save right? I'm not sure what costs more though, doing this once per source or `entry.getValue().getValue().forEachRemaining(values::add)` appending a `List<V>` even if we might filter the key group here?

I wonder if we can make `Boolean acceptKeyGroup = null` and change #410 to something like this:

```java
if (bucketsPerSource.get(tupleTag) >= parallelism || acceptKeyGroup) {
  // source # buckets >= parallelism = definitely keep
  // else, unsure, check hash
  entry.getValue().getValue().forEachRemaining(values::add);
  keyGroupCount += values.size();
} else if (keyGroupFilter.apply(minKeyEntry.getValue().getKey())) {
  acceptKeyGroup = true;
  entry.getValue().getValue().forEachRemaining(values::add);
  keyGroupCount += values.size();
} else {
  // skip key but still have to exhaust iterator
  entry.getValue().getValue().forEachRemaining(value -> {});
}
```

Basically, add items if numBuckets >= parallelism, otherwise lazily check hash, skip key if both are false. WDYT?

Reply:
that's a good idea -- should work. I'll update the PR!